src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,699 @@
+// only consider edges to other private types.
+struct TypeSCCs
+  : SCCs<typename std::vector<HeapType>::const_iterator, TypeSCCs> {
+  std::unordered_set<HeapType> includedTypes;
+  TypeSCCs(const std::vector<HeapType>& types)
+    : SCCs(types.begin(), types.end()),
+      includedTypes(types.cbegin(), types.cend()) {}
+  void pushChildren(HeapType parent) {
+    for (auto child : parent.getReferencedHeapTypes()) {
+      if (includedTypes.count(child)) {
+        push(child);
+      }
+    }
+  }
+};
+
+// After all their distinct permutations have been used, different groups with
+// the same shapes must be differentiated by adding in a "brand" type. Even with
+// a brand mixed in, we might run out of permutations, in which case we need a
+// new brand type. This iterator provides an infinite sequence of possible brand
+// types, prioritizing those with the most compact encoding.
+struct BrandTypeIterator {
+  static constexpr size_t optionCount = 18;
+  static std::array<Field, optionCount> fieldOptions;
+  static void initFieldOptions();
+
+  struct FieldInfo {
+    uint8_t index = 0;
+    bool immutable = false;
+
+    operator Field() const {
+      auto field = fieldOptions[index];
+      if (immutable) {
+        field.mutable_ = Immutable;
+      }
+      return field;
+    }
+
+    bool advance() {
+      if (!immutable) {
+        immutable = true;
+        return true;
+      }
+      immutable = false;
+      index = (index + 1) % optionCount;
+      return index != 0;
+    }
+  };
+
+  bool useArray = false;
+  std::vector<FieldInfo> fields;
+
+  HeapType operator*() const {
+    if (useArray) {
+      return Array(fields[0]);
+    }
+    return Struct(std::vector<Field>(fields.begin(), fields.end()));
+  }
+
+  BrandTypeIterator& operator++() {
+    for (size_t i = fields.size(); i > 0; --i) {
+      if (fields[i - 1].advance()) {
+        return *this;
+      }
+    }
+    if (useArray) {
+      useArray = false;
+      return *this;
+    }
+    fields.emplace_back();
+    useArray = fields.size() == 1;
+    return *this;
+  }
+};
+
+// Create an adjacency list with edges from supertype to subtypes.
+std::vector<std::vector<size_t>>
+createSubtypeGraph(const std::vector<HeapType>& types) {
+  std::unordered_map<HeapType, size_t> indices;
+  for (auto type : types) {
+    indices.insert({type, indices.size()});
+  }
+
+  std::vector<std::vector<size_t>> subtypeGraph(types.size());
+  for (size_t i = 0; i < types.size(); ++i) {
+    if (auto super = types[i].getDeclaredSuperType()) {
+      if (auto it = indices.find(*super); it != indices.end()) {
+        subtypeGraph[it->second].push_back(i);
+      }
+    }
+  }
+  return subtypeGraph;
+}
+
+struct RecGroupInfo;
+
+// As we iterate through the strongly connected components, we may find
+// components that have the same shape. When we find such a collision, we merge
+// the components into a single equivalence class where we track how we have
```

Comment:
```suggestion
// the components into a single group class where we track how we have
```
(or maybe both? either way getting "group" into the comment would help I think)